### PR TITLE
EVP demo for XOF digest using SHAKE256

### LINKED ISSUE #14106: Add an EVP XOF digest demo
Suggestion: SHAKE256 but not compulsory.

It should be the current best practice (fetch), have proper error handling and resource allocation/deallocation and be in the style of a known answer test.

### PR DESCRIPTION
**Pull request by hlandau**
This demo optionally accepts a single command line argument, allowing the output length to be specified.

Fixes #14106.

### CODE REVIEW

**Review by mattcaswell — CHANGES_REQUESTED** [review]
LGTM. A couple of nits

File: demos/digest/EVP_MD_xof.c [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+static const char *propq = NULL;
+
+int main(int argc, char **argv)
+{
+    int rv = 1;
+    OSSL_LIB_CTX *libctx = NULL;
+    EVP_MD *md = NULL;
+    EVP_MD_CTX *ctx = NULL;
+    OSSL_PARAM params[2], *p = params;
+    unsigned digest_len = 20;
```
[the remark this answers is not in the document]
Comment (hlandau) [reply demos/digest/EVP_MD_xof.c]: Noted for future reference. Thanks.

**Comment by hlandau** [comment]
Updated.

**Review by mattcaswell — APPROVED** [review]

File: demos/digest/EVP_MD_xof.c [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+    if (digest == NULL) {
+        fprintf(stderr, "Failed to allocate memory for digest\n");
+        goto end;
+    }
+
+    /* Get computed digest. */
+    if (EVP_DigestFinal(ctx, digest, &digest_len) == 0) {
+        fprintf(stderr, "Failed to finalize hash\n");
+        goto end;
+    }
```
Comment (t8m) [inline demos/digest/EVP_MD_xof.c:121]: I think setting the XOFLEN and using DigestFinal is the less common way how to generate the XOF output. There is also the EVP_DigestFinalXOF() which allows setting the xof length directly as a parameter.

In future (for implementations that allow this) the EVP_DigestFinalXOF() should even allow multiple calls to produce further XOF bytes.

Comment (hlandau) [reply demos/digest/EVP_MD_xof.c]: Updated, TIL.

Out of curiosity, what if we ever have to support a XOF which needs the output length to be specified up front (for example, if the length is placed in an internal tweak)?

**Comment by hlandau** [comment]
Updated.

File: demos/digest/EVP_MD_xof.c (hunk 2 above)
Comment (t8m) [reply demos/digest/EVP_MD_xof.c]: It is not a true XOF then. That's just a variable length hash function. But yeah you'd use the way you wrote the demo originally.

**Review by t8m — APPROVED** [review]

**Comment by t8m** [comment]
@mattcaswell please reapprove.

**Review by mattcaswell — APPROVED** [review]

**Review by paulidale — APPROVED** [review]

**Comment by t8m** [comment]
Merged to master and 3.0 branches. Thank you.

**Pull request closed without merging**